File: frontend/types/map-plan.ts
export type UserProfile = {
  name: string;
  phone: string;
  neighborhood?: string;
  cuisines_loved?: string[];
  cuisines_avoided?: string[];
  dietary_restrictions?: string[];
  price_range?: string;
  availability?: string[];
  preferred_meal_times?: Record<string, string>;
  buffer_mins?: number;
  flexibility_mins?: number;
  transport_mode?: string;
  max_travel_mins?: number;
  vibes?: string[];
  photo_spots?: boolean;
  past_places?: Array<Record<string, unknown>>;
  vetoed_places?: string[];
};

export type MapStartVenue = {
  name: string;
  address: string;
  emoji: string;
  x: number;
  y: number;
};

export type MapOption = {
  id: string;
  name: string;
  address: string;
  phone?: string | null;
  score: number;
  price: string;
  walk?: string | null;
  transit?: string | null;
  vibes: string[];
  dietary?: string | null;
  why: string;
  color: string;
  x: number;
  y: number;
  /** Real-world coordinates for the MapLibre satellite map */
  lat?: number;
  lng?: number;
  ghost: boolean;
  photos: string[];
  reels: string[];
};

export type MapPlanStep = {
  step: number;
  time: string;
  label: string;
  type: string;
  venue?: MapStartVenue | null;
  options: MapOption[];
};

export type BuildMapPlanResponse = {
  group_id: string;
  group: string[];
  neighborhood: string;
  steps: MapPlanStep[];
};

export type ChooseMapPlanOptionResponse = {
  group_id: string;
  choices: Record<string, string>;
  selected_step: number;
  selected_option: MapOption;
  next_step?: MapPlanStep | null;
  final_steps: MapPlanStep[];
  completed: boolean;
};

File: backend/app/schemas/map_plan.py
from pydantic import BaseModel, Field

from app.models.user import UserProfile

class MapStartVenueResponse(BaseModel):
    name: str
    address: str
    emoji: str
    x: int
    y: int


class MapOptionResponse(BaseModel):
    id: str
    name: str
    address: str
    score: int
    price: str
    walk: str | None = None
    transit: str | None = None
    vibes: list[str] = Field(default_factory=list)
    dietary: str | None = None
    why: str
    color: str
    x: int
    y: int
    ghost: bool = False
    photos: list[str] = Field(default_factory=list)
    reels: list[str] = Field(default_factory=list)


class MapPlanStepResponse(BaseModel):
    step: int
    time: str
    label: str
    type: str
    venue: MapStartVenueResponse | None = None
    options: list[MapOptionResponse] = Field(default_factory=list)


class BuildMapPlanRequest(BaseModel):
    group_id: str
    profiles: list[UserProfile]


class BuildMapPlanResponse(BaseModel):
    group_id: str
    group: list[str]
    neighborhood: str
    steps: list[MapPlanStepResponse]


class ChooseMapPlanOptionRequest(BaseModel):
    group_id: str
    profiles: list[UserProfile]
    current_step: int
    selected_option_id: str
    choices: dict[str, str] = Field(default_factory=dict)


class ChooseMapPlanOptionResponse(BaseModel):
    group_id: str
    choices: dict[str, str]
    selected_step: int
    selected_option: MapOptionResponse
    next_step: MapPlanStepResponse | None = None
    final_steps: list[MapPlanStepResponse] = Field(default_factory=list)
    completed: bool = False